### PR TITLE
Adds Axis Object - Refactors Amcharts Axis PPs

### PR DESCRIPTION
**Pull request by mlzummo**
@GordonSmith @jbrundage @dtsnell4 - Please Review

(still need to clean up commented code but would like you to read the commented sections and get ideas for comments)
Signed-off-by: Mathew Zummo <mzummo@gmail.com>

### CODE REVIEW

File: src/amchart/CommonSerial.js [hunk 1]
```diff
@@ -127,8 +78,109 @@
         return retVal;
     };
 
+    //Axes
+    // perhaps look at creating a base Axis/Graph widget in common folder and then the below functions would be in widget.js and we could then override for each widget?
+    var xAxes = CommonSerial.prototype.xAxes;
+    var yAxes = CommonSerial.prototype.yAxes;
+    CommonSerial.prototype.Axes = function (type, _) {
```
Comment (mlzummo) [inline src/amchart/CommonSerial.js]: change _ to idx?

File: src/amchart/CommonXY.js [hunk 2]
```diff
@@ -105,55 +101,64 @@
...
-        this._chart.valueAxes[1].gridAlpha = this.yAxisGridAlpha();
-        this._chart.valueAxes[1].dashLength = this.yAxisDashLength();
-        this._chart.valueAxes[1].axisTitleOffset = this.yAxisTitleOffset();
-
-        this._chart.valueAxes[1].labelFunction = function(d) {
-            return d3.format(context.yAxisTickFormat())(d);
-        };
+        for (var iy = 1; iy < this.yAxes().length; iy+=2) {
+            var yAxis = this.yAxis()[iy];
+            this._chart.valueAxes[iy].position = "left";
```
Comment (mlzummo) [inline src/amchart/CommonXY.js]: need to add a PP for axis position

**Comment by mlzummo** [comment]
if this done become the preferred method .. or something similar would be best to refactor all charts at the same times

File: src/amchart/Axis.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    //Axis.prototype.publish("axisMinimum", null, "number", "",null,{tags:["Intermediate"]});
+    Axis.prototype.publish("axisTitleOffset", null, "number", "",null,{tags:["Intermediate"]});
+
+    Axis.prototype.publish("startOnAxis", true, "boolean", "Draw Chart Starting On Axis.",null,{tags:["Intermediate"]});
+
+    Axis.prototype.publish("axisTypeTimePattern", "%Y-%m-%d", "string", "Time Series Pattern");
+
+    Axis.prototype.publish("axisType", "ordinal", "set", "X-Axis Type", ["ordinal", "linear", "time", "pow", "log", "none"]);
+
+    Axis.prototype.publish("axisTickFormat", "s", "string", "Y-Axis Tick Format");
```
Comment (GordonSmith) [inline src/amchart/Axis.js]: remove blank lines between each publish item.

File: src/amchart/CommonSerial.js (hunk 1 above)
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: yup

File: src/amchart/CommonSerial.js [hunk 4]
```diff
@@ -127,8 +78,109 @@
...
+    //         case "x":
+    //             return this.xAxis(0);
+    //         break;
+    //         case "y":
+    //             return this.yAxis(0);
+    //         default:
+    //             // TODO
+    //     }
+    // }
+
```
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: remove commented code

File: src/amchart/CommonSerial.js [hunk 5]
```diff
@@ -24,6 +23,10 @@
     CommonSerial.prototype.constructor = CommonSerial;
     CommonSerial.prototype._class += " amchart_CommonSerial";
 
+    CommonSerial.prototype.publish("xAxes", [], "widgetArray", "widgets", null, { max: 1, tags: ["Basic"] }); // max number of xAxes
+    CommonSerial.prototype.publish("yAxes", [], "widgetArray", "widgets", null, { tags: ["Basic"] });
+    //CommonSerial.prototype.publish("graphs", [""], "widgetArray", "widgets", null, { tags: ["Basic"] }); // this is breaking serialization why?
```
Comment (jbrundage) [inline src/amchart/CommonSerial.js]: It's possibly breaking serialization because "" isn't a widget. Did you try defaulting to [] ?

**Comment by mlzummo** [comment]
going to submit new PR

**Comment by GordonSmith** [comment]
Try not to submit a new PR - keep this one going by simply updating your current branch, then we won't lose the comment history.

**Comment by mlzummo** [comment]
@GordonSmith @dtsnell4 @jbrundage Please Review - (this is based off some of the amcharts fixes i did as well)

File: src/amchart/Column.js [hunk 6]
```diff
@@ -23,7 +23,7 @@
     Column.prototype.publish("stacked", false, "boolean", "Stack Chart",null,{tags:["Basic","Shared"]});
     Column.prototype.publish("fillOpacity", 0.7, "number", "Opacity of The Fill Color", null, {min:0,max:1,step:0.001,inputType:"range",tags:["Intermediate","Shared"]});
 
-    Column.prototype.publish("paletteGrouping", "By Column", "set", "Palette Grouping",["By Category","By Column"],{tags:["Basic"]});
+    //Column.prototype.publish("paletteGrouping", "By Column", "set", "Palette Grouping",["By Category","By Column"],{tags:["Basic"]});
```
Comment (dtsnell4) [inline src/amchart/Column.js]: comment

Comment (mlzummo) [inline src/amchart/Column.js]: not sure if im going to remove it or not from here yet ....

**Comment by mlzummo** [comment]
@GordonSmith @dtsnell4 @jbrundage - Please Review i think this will be ready to pull soon

**Comment by mlzummo** [comment]
the code technically will support unlimited "y" axis ... idk if amcharts support multiple "x" axes i dont think it does ... but for now we will just expost "y2" and scatter which it supports it has the feature off for now ... but still uses the same axis object stuff ....

**Comment by mlzummo** [comment]
not sure why serialization doesnt work right now

**Comment by mlzummo** [comment]
http://rawgit.com/mlzummo/Visualization/AMCHART_AXIS_REFACTOR/demos/dermatology.html?src/amchart/Column.multiY

File: src/amchart/CommonSerial.js [hunk 7]
```diff
@@ -26,13 +25,16 @@
...
-
     }
     CommonSerial.prototype = Object.create(HTMLWidget.prototype);
 
     CommonSerial.prototype.constructor = CommonSerial;
     CommonSerial.prototype._class += " amchart_CommonSerial";
 
+    CommonSerial.prototype.publish("xAxes", [], "widgetArray", "widgets", null, { max: 1, tags: ["Basic"] }); // max number of xAxes
+    CommonSerial.prototype.publish("yAxes", [], "widgetArray", "widgets", null, { tags: ["Basic"] });
+    //CommonSerial.prototype.publish("graphs", [""], "widgetArray", "widgets", null, { tags: ["Basic"] }); // potentially to do
```
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: remove commented code.

File: src/amchart/CommonSerial.js [hunk 8]
```diff
@@ -131,28 +79,52 @@
 
     CommonSerial.prototype.publish("paletteGrouping", "By Column", "set", "Palette Grouping",["By Category","By Column"],{tags:["Basic"]});
 
-    var xAxisTypeTimePattern = CommonSerial.prototype.xAxisTypeTimePattern;
-    CommonSerial.prototype.xAxisTypeTimePattern = function (_) {
-        var retVal = xAxisTypeTimePattern.apply(this, arguments);
-        if (arguments.length) {
-            this._dateParserData = d3.time.format(_).parse;
+    //CommonSerial.prototype.publish("axeMap", null, "object", "Axis Data Mapping",null,{tags:["Intermediate","Shared"]}); // TODO for 2+ axes
```
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: remove commented code

**Comment by GordonSmith** [comment]
Looks alright - I am not following what your doing setting the context within the individual axes, but if your tests are working I can pull once the commented code is removed etc.?

**Comment by mlzummo** [comment]
@GordonSmith removed comments .... have you tested this yourself ... these are some pretty big changes

**Comment by GordonSmith** [comment]
@dtsnell4 Please review and test...

**Comment by mlzummo** [comment]
remove axistitleoffset

**Comment by jbrundage** [comment]
Looks good to me... works well

**Comment by dtsnell4** [comment]
Looks good to me as well

**Comment by mlzummo** [comment]
@GordonSmith are we ready to pull this?

**Pull request merged**